lib/helpers.js

```diff
@@ -0,0 +1,295 @@
+  if(idx === -1) {
+    return null;
+  }
+
+  // valid root zcap invocation targets:
+  // `/kms/keystores/<keystoreId>/keys`
+  // `/kms/keystores/<keystoreId>/authorizations`
+  // `/kms/keystores/<keystoreId>/revocations`
+  const path = url.substr(idx + 6 /* 'zcaps/'.length */);
+  if(!['keys', 'authorizations', 'revocations'].includes(path)) {
+    return null;
+  }
+
+  // return invocation target for the given root zcap URL
+  return {
+    target: `${keystoreId}/${path}`,
+    keystoreId
+  };
+};
+
+exports.getKeystoreId = ({host, req, localId, routes}) => {
+  if(!host) {
+    host = req.get('host');
+  }
+  return `https://${host}${routes.keystores}/${localId}`;
+};
+
+exports.inspectCapabilityChain = async ({
+  capabilityChain, capabilityChainMeta
+}) => {
+  // collect the capability IDs and delegators for the capabilities in the chain
+  const capabilities = [];
+  for(const [i, capability] of capabilityChain.entries()) {
+    const [{purposeResult}] = capabilityChainMeta[i].verifyResult.results;
+    if(purposeResult && purposeResult.delegator) {
+      capabilities.push({
+        capabilityId: capability.id,
+        delegator: purposeResult.delegator.id,
+      });
+    }
+  }
+  const revoked = await brZCapStorage.revocations.isRevoked({capabilities});
+
+  if(revoked) {
+    return {
+      valid: false,
+      error: new Error(
+        'One or more capabilities in the chain have been revoked.')
+    };
+  }
+
+  return {valid: true};
+};
+
+exports.verifyDelegation = async ({keystoreId, host, capability}) => {
+  const invocationTarget = typeof capability.invocationTarget === 'string' ?
+    capability.invocationTarget : capability.invocationTarget.id;
+
+  const expectedRootCapability = _getExpectedRootCapability(
+    {invocationTarget, keystoreId});
+  const {verified, error, results} = await jsigs.verify(capability, {
+    suite: new Ed25519Signature2018(),
+    purpose: new CapabilityDelegation({
+      expectedRootCapability,
+      inspectCapabilityChain: exports.inspectCapabilityChain,
+      suite: new Ed25519Signature2018()
+    }),
+    documentLoader: _createDocumentLoader({host, keystoreId}),
+    compactProof: false
+  });
+  if(!verified) {
+    throw error;
+  }
+  return results;
+};
+
+/**
+ * Gets an `expectedRootCapability` from an `invocationTarget`. This function
+ * handles special cases where the `invocationTarget` does not match the
+ * `expectedRootCapability` because the `invocationTarget` cannot be expressed
+ * as a zcap.
+ *
+ * @param {object} options - Options to use.
+ * @param {string} options.invocationTarget - An invocationTarget.
+ * @param {string} options.keystoreId - A url to a key store.
+ *
+ * @return {string} The `expectedRootCapability` for the `invocationTarget`.
+ */
+function _getExpectedRootCapability({invocationTarget, keystoreId}) {
+  // `authorizations` endpoint cannot be expressed as a zcap, map to zcap space
+  if(invocationTarget === `${keystoreId}/authorizations`) {
```

Comment:
Need `${keystoreId}/revocations` conditional too.

Reply:
I'm confused about how this is working in the test (which is passing).  Logging tells me that in our existing tests, the only `invocationTarget` value that passes through this code is `https://bedrock.localhost:18443/kms/keystores/z19qzd9odBkJdAtqsnqoxE44o/keys/z1AC8SrmLbTWkyE11SqCAQcyN`.  Nothing with a `revocations` suffix.  

Although we do have such an `invocationTarget` in the test: 

https://github.com/digitalbazaar/bedrock-kms-http/blob/725a56bd26973817ddc6c82cae51a465ee63ca3d/test/mocha/20-revocation.js#L148-L151

Is the test incorrect in some way?

Reply:
I think it's just a missing test (or missing part of the test). We don't revoke a zcap that allows a party to revoke (we don't ever test revoking something like `bobRevocationZcap`.